lib/wallet/src/Cardano/Wallet/Transaction.hs

```diff
@@ -226,15 +231,54 @@ data TransactionCtx = TransactionCtx
     -- ^ Script template regulating delegation credentials
     , txNativeScriptInputs :: Map TxIn (Script KeyHash)
     -- ^ A map of script hashes related to inputs. Only for multisig wallets
-    , txCollateralRequirement :: SelectionCollateralRequirement
-    -- ^ The collateral requirement.
     } deriving Generic
 
 -- | Represents a preliminary selection of tx outputs typically made by user.
 newtype PreSelection = PreSelection { outputs :: [TxOut] }
     deriving stock (Generic, Show)
     deriving newtype (Eq)
 
+-- | Represents a balanced selection.
+--
+data SelectionOf change = Selection
+    { inputs :: !(NonEmpty (TxIn, TxOut))
+        -- ^ Selected inputs.
+    , collateral :: ![(TxIn, TxOut)]
+        -- ^ Selected collateral inputs.
+    , outputs :: ![TxOut]
+        -- ^ User-specified outputs
+    , change :: ![change]
+        -- ^ Generated change outputs.
+    , assetsToMint :: !TokenMap
+        -- ^ Assets to mint.
+    , assetsToBurn :: !TokenMap
+        -- ^ Assets to burn.
+    , extraCoinSource :: !Coin
+        -- ^ An extra source of ada.
+    , extraCoinSink :: !Coin
+        -- ^ An extra sink for ada.
+    }
+    deriving (Eq, Generic, Show)
+
+instance NFData change => NFData (SelectionOf change)
+
+-- | Computes the ada surplus of a selection, assuming there is a surplus.
+--
+-- If there is no surplus, this function returns 'Coin 0'.
+--
+selectionDelta :: SelectionOf TxOut -> Coin
+selectionDelta selection = balanceIn <\> balanceOut
+  where
+    balanceIn =
+        extraCoinSource
+        <> F.foldMap (TxOut.coin . snd) inputs
+    balanceOut =
+        extraCoinSink
+        <> F.foldMap TxOut.coin outputs
+        <> F.foldMap TxOut.coin change
+    Selection
+        {inputs, outputs, change, extraCoinSource, extraCoinSink} = selection
+
```

Comment:
Hope we can get rid of this eventually

Reply:
> Hope we can get rid of this eventually

Definitely!

Reply:
https://cardanofoundation.atlassian.net/browse/ADP-3158